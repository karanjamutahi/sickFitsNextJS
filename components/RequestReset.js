import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';

import Form from './styles/Form';
import Error from './ErrorMessage';


const REQUEST_RESET_MUTATION = gql`
    mutation REQUEST_RESET_MUTATION(
        $email: String!,
    ) {
        requestReset(
            email: $email
        ) {
            message
        }
    }
`;

class RequestReset extends Component {
    state = {
        email: '',
        showInputs: true,
    };

    saveToState = (e) => {
        this.setState({ [e.target.name] : e.target.value });
    };

    render() {
        return (
            <Mutation mutation={REQUEST_RESET_MUTATION} variables={{email:this.state.email}}>
                {(requestReset, { loading, error, called }) => {
                return(
                <Form data-test="form" method="POST" onSubmit={ async (e) => {
                    e.preventDefault();
                    const res = await requestReset();
                    this.setState({
                        email: '',
                        link: res.data.requestReset.message,
                    });
                }}>
                    <fieldset disabled={loading} aria-busy={loading}>
                        <h2>Reset Your Password</h2>
                        <Error error={error}></Error>
                        {(!error && !loading && called) ? 
                        <p>We've sent a reset <a href={this.state.link}>link</a> to your email </p> :
                        <>
                        <label htmlFor="email">
                            Email
                            <input type="email" name="email" id="email" placeholder='john.doe@gmail.com' required value={this.state.email} onChange={this.saveToState}/>
                        </label>
                        <button type="submit">Reset Password</button>
                        </>
                        }
                    </fieldset>
                </Form>
                )}}
            </Mutation>
        );
    }
}


export default RequestReset;
export { REQUEST_RESET_MUTATION };