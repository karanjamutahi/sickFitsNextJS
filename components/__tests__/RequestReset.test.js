import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import RequestReset, { REQUEST_RESET_MUTATION } from '../RequestReset';

const mocks = [
    {
        request: { query: REQUEST_RESET_MUTATION, variables: {email: "mutahi.henry@gmail.com"}},
        result: { data: { requestReset: { message: "success", __typename: "Message" } } }
    }
]
describe("RequestReset Component", () => {
    it("Renders and matches snapshot", async () => {
        const wrapper = mount(
        <MockedProvider>
            <RequestReset></RequestReset>
        </MockedProvider>
        );
        const form = wrapper.find('form[data-test="form"]');
        //console.log(form.debug());
        expect(toJSON(form)).toMatchSnapshot();
    });

    it("Calls the mutation", async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <RequestReset></RequestReset>
            </MockedProvider>
        );
        //console.log(wrapper.debug());
        //Simulate typing an email
        wrapper.find('input').simulate('change', { target: { name: 'email', value: 'mutahi.henry@gmail.com' } } );
        wrapper.find('form').simulate('submit');
        await wait();
        wrapper.update();
        expect(wrapper.find('p').text()).toContain("We've sent a reset link to your email");
    });
});