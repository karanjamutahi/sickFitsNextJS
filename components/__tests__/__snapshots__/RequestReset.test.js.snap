// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestReset Component Renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 gvKJrt"
  data-test="form"
  method="POST"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Reset Your Password
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="john.doe@gmail.com"
        required={true}
        type="email"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Reset Password
    </button>
  </fieldset>
</form>
`;
